GVFS/GVFS.Service/GvfsService.cs

```diff
@@ -355,37 +355,103 @@ private void CreateAndConfigureProgramDataDirectories()
             this.serviceDataLocation = Paths.GetServiceDataRoot(this.serviceName);
             string serviceDataRootPath = Path.GetDirectoryName(this.serviceDataLocation);
 
-            DirectorySecurity programDataSecurity = new DirectorySecurity();
+            DirectorySecurity serviceDataRootSecurity;
+            if (Directory.Exists(serviceDataRootPath))
+            {
+                serviceDataRootSecurity = Directory.GetAccessControl(serviceDataRootPath);
+            }
+            else
+            {
+                serviceDataRootSecurity = new DirectorySecurity();
+            }
 
             // Protect the access rules from inheritance
-            programDataSecurity.SetAccessRuleProtection(isProtected: true, preserveInheritance: false);
+            serviceDataRootSecurity.SetAccessRuleProtection(isProtected: true, preserveInheritance: false);
+
+            // Remove any rules already set for the directory
+            AuthorizationRuleCollection currentRules = serviceDataRootSecurity.GetAccessRules(includeExplicit: true, includeInherited: true, targetType: typeof(NTAccount));
+            foreach (AuthorizationRule authorizationRule in currentRules)
+            {
+                FileSystemAccessRule fileSystemRule = authorizationRule as FileSystemAccessRule;
+                if (fileSystemRule != null)
+                {
+                    serviceDataRootSecurity.RemoveAccessRule(fileSystemRule);
+                }
+            }

```

Comment:
There is some code duplication between `WindowsFileSystem.cs` and the lines above. `Service` is trying to create a directory with a set of permissions; wouldn't it be better to create a generic function which does that inside the platform specific `WindowsFileSystem` and make `Service` use it.

Reply:
Good suggestion, I've moved the common code into helper methods.  Once I push the latest changes let me know what you think.